File: libs/shape/src/test/kotlin/mono/shape/collection/TwoWayQuickMapTest.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.collection

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * A test for [TwoWayQuickMap]
 */
class TwoWayQuickMapTest {
    private val target = TwoWayQuickMap<Key, Value>()

    @Test
    fun testAdd() {
        target[K1] = V1
        assertEquals(V1, target[Key("k1")])
        assertEquals(setOf(K1), target.getKeys(Value("v1")).toSet())
        assertNull(target[K2])
        assertEquals(emptySet(), target.getKeys(Value("v2")).toSet())

        target[K2] = V1
        assertEquals(V1, target[Key("k1")])
        assertEquals(V1, target[Key("k2")])
        assertEquals(setOf(K1, K2), target.getKeys(Value("v1")).toSet())
        // Same id but different object
        assertNotEquals(setOf(Key("k1"), Key("k2")), target.getKeys(Value("v1")).toSet())
    }

    @Test
    fun testAdd_Conflict() {
        target[K1] = V1
        target[K1] = V1
        target[K2] = V1
        target[K3] = V2

        target[K1] = V2
        assertEquals(V2, target[K1])
        assertEquals(V1, target[K2])
        assertEquals(V2, target[K3])

        assertEquals(setOf(K1, K3), target.getKeys(V2).toSet())
        assertEquals(setOf(K2), target.getKeys(V1).toSet())
    }

    @Test
    fun testRemoveKey() {
        target.removeKey(K1)

        target[K1] = V1
        target[K2] = V2
        target.removeKey(K1)
        assertNull(target[K1])
        assertTrue(target.getKeys(V1).isEmpty())
        assertEquals(V2, target[K2])
    }

    @Test
    fun testRemoveValue() {
        target.removeValue(V1)

        target[K1] = V1
        target[K2] = V1

        target[K3] = V2

        target.removeValue(V1)
        assertEquals(emptySet(), target.getKeys(V1).toSet())
        assertNull(target[K1])
        assertNull(target[K2])
        assertEquals(V2, target[K3])
        assertEquals(setOf(K3), target.getKeys(V2).toSet())
    }

    @Test
    fun testGetKey() {
        target[K1] = V1

        assertEquals(K1, target.getKey(Key("k1")))
    }

    private class Key(override val id: String) : Identifier

    private class Value(override val id: String) : Identifier

    companion object {
        private val K1 = Key("k1")
        private val K2 = Key("k2")
        private val K3 = Key("k3")

        private val V1 = Value("v1")
        private val V2 = Value("v2")
    }
}

File: monosketch-svelte/src/lib/mono/shape/collection/two-way-quick-map.test.ts
import type { Identifier } from "$mono/shape/collection/identifier";
import { beforeEach, describe, expect, test } from "vitest";
import { TwoWayQuickMap } from './two-way-quick-map'; // Adjust the import path as necessary

class Key implements Identifier {
    id: string;

    constructor(id: string) {
        this.id = id;
    }
}

class Value implements Identifier {
    id: string;

    constructor(id: string) {
        this.id = id;
    }
}

const K1 = new Key('k1');
const K2 = new Key('k2');
const K3 = new Key('k3');
const V1 = new Value('v1');
const V2 = new Value('v2');

describe('TwoWayQuickMap', () => {
    let target: TwoWayQuickMap<Key, Value>;

    beforeEach(() => {
        target = new TwoWayQuickMap<Key, Value>();
    });

    test('add', () => {
        target.set(K1, V1);
        expect(target.get(new Key('k1'))).toEqual(V1);
        expect(Array.from(target.getKeys(new Value('v1')))).toEqual([K1]);
        expect(target.get(K2)).toBeUndefined();
        expect(Array.from(target.getKeys(new Value('v2')))).toEqual([]);

        target.set(K2, V1);
        expect(target.get(new Key('k1'))).toEqual(V1);
        expect(target.get(new Key('k2'))).toEqual(V1);
        expect(Array.from(target.getKeys(new Value('v1')))).toEqual([K1, K2]);
    });

    test('add - conflict', () => {
        target.set(K1, V1);
        target.set(K1, V1);
        target.set(K2, V1);
        target.set(K3, V2);

        target.set(K1, V2);
        expect(target.get(K1)).toEqual(V2);
        expect(target.get(K2)).toEqual(V1);
        expect(target.get(K3)).toEqual(V2);

        expect(Array.from(target.getKeys(V2))).toEqual([K3, K1]);
        expect(Array.from(target.getKeys(V1))).toEqual([K2]);
    });

    test('removeKey', () => {
        target.removeKey(K1);

        target.set(K1, V1);
        target.set(K2, V2);
        target.removeKey(K1);
        expect(target.get(K1)).toBeUndefined();
        expect(Array.from(target.getKeys(V1))).toEqual([]);
        expect(target.get(K2)).toEqual(V2);
    });

    test('removeValue', () => {
        target.removeValue(V1);

        target.set(K1, V1);
        target.set(K2, V1);
        target.set(K3, V2);

        target.removeValue(V1);
        expect(Array.from(target.getKeys(V1))).toEqual([]);
        expect(target.get(K1)).toBeUndefined();
        expect(target.get(K2)).toBeUndefined();
        expect(target.get(K3)).toEqual(V2);
        expect(Array.from(target.getKeys(V2))).toEqual([K3]);
    });

    test('getKey', () => {
        target.set(K1, V1);

        expect(target.getKey(new Key('k1'))).toEqual(K1);
    });
});
